### PR TITLE
pkg/types/config/libvirt/cache: Cache unpacked files as well

### PR DESCRIPTION
**Pull request by wking**
Libvirt can't use the HTTP cache directly, because the cached content contains HTTP headers and other metadata.  Rather than writing our own [HTTP cache][1] to perform this unpacking, I've just added a separate cache layer for the unpacked images.

I'm currently trusting upstream ETags as cache keys, but if we want to be paranoid at the expense of hashing a gigabyte or two of data, we could use SHA1 sums of the payload or similar.

I'm using [`Flock`][2] to protect against parallel calls racing unpacks. My intention is to have:

1. Alice gets the lock, starts unpacking into the image cache.
2. Bob blocks on the lock.
3. Alice finishes unpacking and renames to the final filename.
4. Charlie hits the same ETag and grabs the unpacked image without even checking the lock.
5. Alice releases the lock.
6. Bob aquires the lock, sees that the final image is already in place, and uses it without unpacking again.

The quote-splitting extracts [the opaque-tag component from the entity-tag value][3].  Slashes and other filesystem-sensitive components are also legal opaque-tag value, so I'm washing the opaque-tag through MD5 to guard against that.  The MD5 wash also ensures that the `.tmp` and `.lock` suffixes are safe, since the `.` separator (and none of the other suffix characters) are not valid hex.

We'll still leak cache into `imageCacheDir` (like we used to leak into `/tmp`), but:

* Now it's off of `/tmp`, which can cause memory issues for folks who use a tmpfs for `/tmp` and don't have gigs of extra memory ;).
* Now we only leak one unpacked image per ETag, while before we leaked one unpacked image per non-`file://` `UseCachedImage` call.

CC @staebler

[1]: https://godoc.org/github.com/gregjones/httpcache#Cache
[2]: https://godoc.org/golang.org/x/sys/unix#Flock
[3]: https://tools.ietf.org/html/rfc7232.html#section-2.3

### CODE REVIEW

**Comment by wking** [comment]
[e2e][1]:

```
Waiting for API at https://ci-op-6wbfjmjd-1d3f3-api.origin-ci-int-aws.dev.rhcloud.com:6443 to respond ...
Another process exited
2018/10/02 23:28:15 Container test in pod e2e-aws failed, exit code 1, reason Error
```

/retest

[1]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/395/pull-ci-openshift-installer-master-e2e-aws/345/build-log.txt

**Comment by abhinavdahiya** [comment]
@wking https://godoc.org/github.com/gregjones/httpcache#Cache seems fine to use; it has tests :p .
All the new code around caching we have in the repo doesn't seem to have any tests.

We is using something like https://godoc.org/github.com/gregjones/httpcache#Cache not an option?

File: pkg/types/config/libvirt/cache.go [hunk 1]
```diff
@@ -55,20 +56,98 @@ func (libvirt *Libvirt) UseCachedImage() (err error) {
...
+	opaque := etagSections[1]
+	if opaque == "" {
+		return "", fmt.Errorf("caching is not supported when the opaque tag is unset: %s", etag)
+	}
+	hashed := md5.Sum([]byte(opaque))
+	return hex.EncodeToString(hashed[:]), nil
+}
+
+func cacheImage(reader io.Reader, imagePath string) (err error) {
+	logrus.Infof("Unpacking OS image into %q...", imagePath)
```
Comment (crawford) [inline pkg/types/config/libvirt/cache.go]: Can we make this Debugf?

File: pkg/types/config/libvirt/cache.go [hunk 2]
```diff
@@ -55,20 +56,98 @@ func (libvirt *Libvirt) UseCachedImage() (err error) {
...
+	imageCacheDir := filepath.Join(cacheDir, "image")
+	err = os.MkdirAll(imageCacheDir, 0777)
+	if err != nil {
+		return err
+	}
+
+	imagePath := filepath.Join(imageCacheDir, key)
+	_, err = os.Stat(imagePath)
+	if err == nil {
+		logrus.Infof("Using cached OS image %q", imagePath)
```
Comment (crawford) [inline pkg/types/config/libvirt/cache.go]: Can we make this Debugf?

File: pkg/types/config/libvirt/cache.go (hunk 1 above)
Comment (wking) [reply pkg/types/config/libvirt/cache.go]: > Can we make this Debugf?

Done in e62270c -> 6625737.

File: pkg/types/config/libvirt/cache.go (hunk 2 above)
Comment (wking) [reply pkg/types/config/libvirt/cache.go]: > Can we make this Debugf?

Done in e62270c -> 6625737.

**Pull request merged**